src/unity/toolkits/neural_net/coreml_import.cpp

```diff
@@ -0,0 +1,201 @@
+  }
+
+  const float* data() const override { return weights_.floatvalue().data(); }
+  size_t size() const override { return weights_.floatvalue_size(); }
+  const size_t* shape() const override { return shape_.data(); }
+  size_t dim() const override { return shape_.size(); }
+
+private:
+  std::vector<size_t> shape_;
+  WeightParams weights_;
+};
+
+// The overloaded extract_network_params functions below traverse a CoreML spec
+// proto recursively, destructively moving the WeightParams values it finds
+// inside of neural networks into an output float_array_map.
+
+void extract_network_params(const std::string& name,
+                            ConvolutionLayerParams* convolution,
+                            float_array_map* params_out) {
+  ASSERT_EQ(2, convolution->kernelsize_size());
+  const size_t n = convolution->outputchannels();
+  const size_t c = convolution->kernelchannels();
+  const size_t h = convolution->kernelsize(0);
+  const size_t w = convolution->kernelsize(1);
+
+  shared_float_array weights = weight_params_float_array::create_shared(
+      {n, c, h, w}, convolution->mutable_weights());
+  params_out->emplace(name + "_weight", std::move(weights));
+
+  if (convolution->has_bias()) {
+    shared_float_array bias = weight_params_float_array::create_shared(
+        {n}, convolution->mutable_bias());
+    params_out->emplace(name + "_bias", std::move(bias));
+  }
+}
+
+void extract_network_params(const std::string& name,
+                            BatchnormLayerParams* batch_norm,
+                            float_array_map* params_out) {
+  const size_t n = batch_norm->channels();
+
+  shared_float_array gamma = weight_params_float_array::create_shared(
+      {n}, batch_norm->mutable_gamma());
+  params_out->emplace(name + "_gamma", std::move(gamma));
+
+  shared_float_array beta = weight_params_float_array::create_shared(
+      {n}, batch_norm->mutable_beta());
+  params_out->emplace(name + "_beta", std::move(beta));
+
+  shared_float_array mean = weight_params_float_array::create_shared(
+      {n}, batch_norm->mutable_mean());
+  params_out->emplace(name + "_running_mean", std::move(mean));
+
+  shared_float_array variance = weight_params_float_array::create_shared(
+      {n}, batch_norm->mutable_variance());
+  params_out->emplace(name + "_running_var", std::move(variance));
+}
+
+void extract_network_params(NeuralNetworkLayer* neural_net_layer,
+                            float_array_map* params_out) {
+  switch (neural_net_layer->layer_case()) {
+  case NeuralNetworkLayer::kConvolution:
+    extract_network_params(neural_net_layer->name(),
+                           neural_net_layer->mutable_convolution(), params_out);
+    break;
+  case NeuralNetworkLayer::kBatchnorm:
+    extract_network_params(neural_net_layer->name(),
+                           neural_net_layer->mutable_batchnorm(), params_out);
+    break;
+  default:
+    break;
```

Comment:
No, actually: this code is written as basically a recursive traversal of the proto, extracting/dumping any supported layer parameters that it finds. It's true that the current implementation is specific to the object detection network, for which we only care about convolution and batch-norm layers, but even then it's simplest just to allow the recursive traversal to hit every `NeuralNetworkLayer` embedded inside the top-level `Model`, ignoring the unsupported layers. I'm hoping that we can just augment this code to support e.g. style transfer down the road, although ultimately we may want something that "converts" a CoreML model instead of what this does, which is closer to "dumping the embedded parameters, keyed by layer name".